Extend SCRAM auth - SHA-224, SHA-384, SHA-512 (#2713)
esl/MongooseIM: doc/migrations/3.6.0_3.6.x.md

## Extended SCRAM-SHA Support

Since this version, SCRAM authentication mechanisms were extended to support additional hashing algorithms.
So far only SHA-1 was available for hashing and now SHA-224, SHA-256, SHA-384 and SHA-512 are also supported.
This includes the authentication mechanisms and the password format that is stored.

> Please note that enabling and using this functionality might require to adjust the server setup.

Reviewer: ```suggestion
Please note that enabling and using this functionality might require adjusting the server setup.
```

### SASL mechanisms

The possible list of allowed SALS mechanisms was changed. We've added new and more secure methods that can be used during stream negotiation.

Please note that if you were using the following in the configurations file

`{sasl_mechanisms, [cyrsasl_scram]}`

> Using `cyrsasl_scram` as `sasl_mechanism` is now incorrect.

Reviewer: ```suggestion
using `cyrsasl_scram` as `sasl_mechanism` is now incorrect.
```

You can achieve the same result of allowing the usage of SHA-1 with SCRAM authentication mechanism with:

`{sasl_mechanisms, [cyrsasl_scram_sha1]}`

You can also specify a list of all supported SCRAM-SHA mechanisms with:

`{sasl_mechanisms, [cyrsasl_scram_sha1, cyrsasl_scram_sha224, cyrsasl_scram_sha256, cyrsasl_scram_sha384, cyrsasl_scram_sha512]}`

> Before setting up this configuration, please make sure that the client application is capable to authenticate with selected set of authentication mechanisms.

Reviewer: ```suggestion
Before setting up this configuration, please make sure that the client application is capable of authenticating with a selected set of authentication mechanisms.
```

> For more details please refer to [authentication](../../Advanced-configuration#authentication) section.

Reviewer: ```suggestion
For more details please refer to the [authentication](../../Advanced-configuration#authentication) section.
```

### SCRAM password format

> To complement the extensions of the authentication mechanisms, SCRAM password format was also updated.

Reviewer: ```suggestion
To complement the extensions of the authentication mechanisms, the SCRAM password format was also updated.
```

Legacy plaintext and SCRAM formats are still supported.

> Nonetheless, please note that if you were using SCRAM as a password format, this meant that SHA-1 was used as hashing algorithm.

Reviewer: ```suggestion
Nonetheless, please note that if you were using SCRAM as a password format, this meant that SHA-1 was used as the hashing algorithm.
```

> This allowed to authenticate with PLAINTEXT and SCRAM-SHA-1.

Reviewer: ```suggestion
This allowed authenticating with PLAINTEXT and SCRAM-SHA-1.
```

In the new setup the user will still authenticate with those mechanisms given the possible slight syntax change explained [above](#sasl-mechanisms).

However, mixing of the old password format with the new authentication mechanisms can lead to conflicting situations where:

1. A user wants to authenticate with e.g. SCRAM-SHA-256.
2. His old password format is only storing SHA-1 password hash.
3. The authentication fails as it is not possible to derive SHA-256 hash from SHA-1.

> If you want to use the new password format with full set of supported SHA hashes, a password change is required which will calculate all the new SHA hashes.

Reviewer: ```suggestion
If you want to use the new password format with a full set of supported SHA hashes, a password change is required to calculate all the new SHA hashes.
```

Author: Thanks @goddammit for the quick review. Much appreciated!
